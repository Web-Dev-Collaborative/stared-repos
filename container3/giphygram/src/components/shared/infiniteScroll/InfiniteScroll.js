import React from 'react';
import PropTypes from 'prop-types';

const DEFAULT_ACTIVATION_DISTANCE = 50;

export class InfiniteScroll extends React.Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node,
    ]).isRequired,
    onFetchMore: PropTypes.func,
    activationDistance: PropTypes.number,
  };

  static defaultProps = {
    onFetchMore: () => {},
    activationDistance: DEFAULT_ACTIVATION_DISTANCE,
  };

  componentDidMount() {
    window.addEventListener('scroll', this.onScroll, false);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.onScroll, false);
  }

  onScroll = () => {
    const { activationDistance, onFetchMore } = this.props;

    const documentHeight = document.body.offsetHeight;
    const scrollHeight = window.innerHeight + window.scrollY;

    if ((scrollHeight + activationDistance) >= documentHeight) {
      onFetchMore();
    }
  };

  render() {
    const { children } = this.props;

    return (
      <div>
        {children}
      </div>
    );
  }
}
