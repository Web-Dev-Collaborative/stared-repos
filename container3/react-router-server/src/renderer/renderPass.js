import React from 'react';
import { renderToString, renderToStaticMarkup } from 'react-dom/server';
import AsyncRenderer from '../components/AsyncRenderer';
import removeDuplicateModules from '../utils/removeDuplicateModules';

const renderPass = (context, element, staticMarkup = false) => {
  context.callback = () => {
    if (context.finishedLoadingModules && !context.statesRenderPass) {
      context.statesRenderPass = true;
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    } else if (context.finishedLoadingModules && context.statesRenderPass || !context.hasModules) {
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    }
  };

  let component = (
    <AsyncRenderer context={context}>
      {element}
    </AsyncRenderer>
  );

  let result;
  try {
    if (staticMarkup) {
      result = renderToStaticMarkup(component);
    } else {
      result = renderToString(component);
    }
  } catch (e) {
    return context.reject(e);
  }

  if (!context.hasModules && !context.hasStates) {
    context.resolve({
      html: result,
      state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
      modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
    });
  }
  return result;
};

export default renderPass;
