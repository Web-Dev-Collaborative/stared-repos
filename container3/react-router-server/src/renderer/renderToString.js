import renderPass from './renderPass';
import { clear } from '../module/cache';

export default (element) => {
  return new Promise((resolve, reject) => {
    clear();
    const context = {
      resolve, reject,
      modulesLoading: 0,
      fetchingStates: 0
    };
    renderPass(context, element);
  });
};
