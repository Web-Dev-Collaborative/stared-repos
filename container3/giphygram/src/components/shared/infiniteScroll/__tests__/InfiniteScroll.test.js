import React from 'react';
import renderer from 'react-test-renderer';
import { InfiniteScroll } from '../InfiniteScroll';

// Create a spy on window.addEventListener() and removeEventListener() functions.
global.addEventListener = jest.fn();
global.removeEventListener = jest.fn();

describe('InfiniteScroll', () => {
  // We need scrollCallback to imitate onScroll window event.
  let onScrollCallback;
  const onScrollCallbackRemove = jest.fn();

  beforeEach(() => {
    // Mock window.addEventListener implementation to be able to fire onScroll callbacks.
    global.addEventListener.mockImplementation((eventName, callback, useCapture) => {
      onScrollCallback = callback;
    });

    // Mock window.addEventListener implementation to be able to fire onScroll callbacks.
    global.removeEventListener.mockImplementation((eventName, callback, useCapture) => {
      if (eventName === 'scroll') {
        onScrollCallbackRemove();
      }
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render one child component correctly', () => {
    const tree = renderer
      .create((
        <InfiniteScroll>
          <div>Child #1</div>
        </InfiniteScroll>
      ))
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render many children components correctly', () => {
    const tree = renderer
      .create((
        <InfiniteScroll>
          <div>Child #1</div>
          <div>Child #2</div>
        </InfiniteScroll>
      ))
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should fire onFetchMore callback', () => {
    // Mock onFetchMore function.
    const onFetchMore = jest.fn();

    // Let's render component.
    renderer.create((
      <InfiniteScroll onFetchMore={onFetchMore} activationDistance={0}>
        <div>Child</div>
      </InfiniteScroll>
    ));

    // Check callbacks after component is rendered.
    expect(onScrollCallback).toBeDefined();
    expect(onFetchMore).not.toHaveBeenCalled();

    // Imitate window onScroll event.
    onScrollCallback();
    expect(onFetchMore).toHaveBeenCalledTimes(1);

    // Try to reduce inner height to prevent onScroll callback from calling.
    // @TODO: Try to set up real values for document.body.offsetHeight.
    // document.body.offsetHeight = 42;
    global.innerHeight = -1;
    global.scrollY = 0;

    onScrollCallback();
    expect(onFetchMore).toHaveBeenCalledTimes(1);
  });

  it('should remove onScroll listener', () => {
    const testComponent = renderer.create((
      <InfiniteScroll>
        <div>Child</div>
      </InfiniteScroll>
    ));

    expect(global.addEventListener).toHaveBeenCalled();
    expect(onScrollCallbackRemove).not.toHaveBeenCalled();

    testComponent.unmount();

    expect(onScrollCallbackRemove).toHaveBeenCalled();
  });
});
