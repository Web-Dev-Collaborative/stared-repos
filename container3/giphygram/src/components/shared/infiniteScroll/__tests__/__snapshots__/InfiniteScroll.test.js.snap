// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteScroll should render many children components correctly 1`] = `
<div>
  <div>
    Child #1
  </div>
  <div>
    Child #2
  </div>
</div>
`;

exports[`InfiniteScroll should render one child component correctly 1`] = `
<div>
  <div>
    Child #1
  </div>
</div>
`;
